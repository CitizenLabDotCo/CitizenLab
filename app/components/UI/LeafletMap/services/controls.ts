import L from 'leaflet';

export function addLayersControl(map: L.Map) {
  return L.control.layers().addTo(map);
}

export function removeControl(
  map: L.Map,
  layerControl?: L.Control.Layers | null
) {
  if (layerControl) {
    map.removeControl(layerControl);
  }
}
