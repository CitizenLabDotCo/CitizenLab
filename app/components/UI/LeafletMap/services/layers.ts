import {
  isNilOrError,
  isFunction,
  isString,
  isOrReturnsString,
} from 'utils/helperUtils';
import L from 'leaflet';
import { isEmpty, cloneDeep, reverse } from 'lodash-es';

import { markerIcon } from './markers';

import {
  GeoJSONLayer,
  IMarkerStringOrObjectOrFunctionForLayer,
  IOverlayStringOrObjectOrFunctionForLayer,
  ITooltipStringOrObjectOrFunctionForLayer,
  IPopupStringOrObjectOrFunctionForLayer,
} from '../typings';

import { DEFAULT_MARKER_ICON } from '../config';

export function addPopupToLayer(
  layer: L.GeoJSON,
  feature: GeoJSON.Feature,
  popupStringOrFunction?: IPopupStringOrObjectOrFunctionForLayer
) {
  if (!popupStringOrFunction) {
    return;
  }

  let popup: string = '';

  if (isFunction(popupStringOrFunction)) {
    popup = popupStringOrFunction(layer, feature);
  } else if (isString(popupStringOrFunction)) {
    popup = popupStringOrFunction;
  }

  if (popup && !isEmpty(popup)) {
    layer.bindPopup(popup);
  }
}

export function addTooltipToLayer(
  layer: L.GeoJSON,
  feature: GeoJSON.Feature,
  tooltipStringOrFunction?: ITooltipStringOrObjectOrFunctionForLayer
): void {
  if (!tooltipStringOrFunction) {
    return;
  }

  let tooltip: string = '';

  if (isFunction(tooltipStringOrFunction)) {
    tooltip = tooltipStringOrFunction(layer, feature);
  } else if (isString(tooltipStringOrFunction)) {
    tooltip = tooltipStringOrFunction;
  }

  if (tooltip && !isEmpty(tooltip)) {
    layer.bindTooltip(tooltip);
  }
}

export function addLayerOverlay(
  layer: L.GeoJSON,
  layerControl: L.Control.Layers,
  geoJSONLayer: GeoJSONLayer,
  overlayStringOrOptionsOrFunction: IOverlayStringOrObjectOrFunctionForLayer
): void {
  let overlayContent;

  if (isString(overlayStringOrOptionsOrFunction)) {
    overlayContent = overlayStringOrOptionsOrFunction;
  } else if (isFunction(overlayStringOrOptionsOrFunction)) {
    overlayContent = overlayStringOrOptionsOrFunction(geoJSONLayer);
  }

  if (isString(overlayContent)) {
    layerControl.addOverlay(layer, overlayContent);
  }
}

export function layerMarker(
  geojsonLayer: GeoJSONLayer,
  latlng: L.LatLng,
  markerStringOrOptionsOrFunction: IMarkerStringOrObjectOrFunctionForLayer,
  options = {}
): L.Marker {
  let marker: L.Icon;

  if (isString(markerStringOrOptionsOrFunction)) {
    marker = markerIcon({ url: markerStringOrOptionsOrFunction });
  } else if (
    isOrReturnsString(markerStringOrOptionsOrFunction, geojsonLayer, latlng)
  ) {
    marker = markerIcon({
      url: markerStringOrOptionsOrFunction(geojsonLayer, latlng),
    });
  } else if (isFunction(markerStringOrOptionsOrFunction)) {
    marker = markerIcon(markerStringOrOptionsOrFunction(geojsonLayer, latlng));
  } else {
    marker = markerIcon(markerStringOrOptionsOrFunction);
  }

  return L.marker(latlng, { ...options, icon: marker || DEFAULT_MARKER_ICON });
}

export function addLayers(
  map: L.Map,
  layersGeoJson: GeoJSONLayer[],
  { layerControl, overlay, popup, tooltip, marker }
): L.GeoJSON[] {
  return reverse(cloneDeep(layersGeoJson))
    ?.filter((layerObject) => !isEmpty(layerObject.geojson))
    .map((layerObject) => {
      const options = {
        useSimpleStyle: true,
        useMakiMarkers: true,
        pointToLayer: (_feature, latlng) => {
          layerMarker(layerObject, latlng, marker);
        },
        onEachFeature: (feature, layer) => {
          addTooltipToLayer(layer, feature, tooltip);
          addPopupToLayer(layer, feature, popup);
        },
      };

      const layer = L.geoJSON(layerObject.geojson, options as any).addTo(map);

      if (!isNilOrError(layerControl) && !isNilOrError(overlay)) {
        addLayerOverlay(layer, layerControl, layerObject, overlay);
      }

      return layer as L.GeoJSON;
    });
}

export function removeLayers(map: L.Map, leafletLayers: L.Layer[]): void {
  leafletLayers.forEach((layer) => {
    removeLayer(map, layer);
  });
}

export function removeLayer(map: L.Map, leafletLayer: L.Layer): void {
  if (leafletLayer) {
    map.removeLayer(leafletLayer);
  }
}
