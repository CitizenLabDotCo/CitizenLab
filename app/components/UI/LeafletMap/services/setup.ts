import L from 'leaflet';

import {
  DEFAULT_TILE_PROVIDER,
  DEFAULT_TILE_OPTIONS,
  DEFAULT_CENTER,
  DEFAULT_ZOOM,
  DEFAULT_BOUND_CONFIG,
} from '../config';

import { IOnMapClickHandler } from '../typings';

export function setup(
  mapId,
  { tileProvider, onClick, center, zoom, onMoveHandler, onZoomHandler }
) {
  const map = init(mapId);

  addTileLayer(map, tileProvider);
  addOnClickHandler(map, onClick);
  changeView(map, center, zoom);
  addOnMoveHandler(map, onMoveHandler);
  addOnZoomHandler(map, onZoomHandler);

  return map;
}

export function init(mapId: string): L.Map {
  return L.map(mapId);
}

export function addTileLayer(
  map: L.Map,
  tileProvider: string = DEFAULT_TILE_PROVIDER
) {
  if (!map) {
    return;
  }

  L.tileLayer(tileProvider, DEFAULT_TILE_OPTIONS).addTo(map);
}

export function addOnClickHandler(
  map: L.Map,
  onClickHandler?: IOnMapClickHandler
): void {
  if (onClickHandler) {
    map.on('click', (event: L.LeafletMouseEvent) => {
      onClickHandler(map, event.latlng);
    });
  }
}

export function changeView(
  map: L.Map,
  center?: L.LatLngExpression,
  zoom?: number
): void {
  map.setView(
    (center || DEFAULT_CENTER) as L.LatLngExpression,
    zoom || DEFAULT_ZOOM
  );
}

export function addOnMoveHandler(
  map: L.Map,
  onMoveHandler: (latlng: [number, number]) => void
): void {
  map.on('moveend', () => {
    const center = map.getCenter();
    onMoveHandler([center.lat, center.lng]);
  });
}

export function addOnZoomHandler(
  map: L.Map,
  onZoomHandler: (latlng: number) => void
): void {
  map.on('zoomend', () => {
    const zoom = map.getZoom();
    onZoomHandler(zoom);
  });
}

export function refitBounds(
  map: L.Map,
  bounds: L.LatLngBounds,
  options: { fitBounds: boolean } = { fitBounds: true }
) {
  if (options.fitBounds) {
    map.fitBounds(bounds, DEFAULT_BOUND_CONFIG);
  }
}
