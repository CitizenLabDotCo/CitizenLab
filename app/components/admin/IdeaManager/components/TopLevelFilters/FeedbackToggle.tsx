import React from 'react';
import { adopt } from 'react-adopt';
import { isNilOrError } from 'utils/helperUtils';

// i18n
import { FormattedMessage } from 'utils/cl-intl';
import messages from '../../messages';

// styles
import styled, { css } from 'styled-components';
import { colors, fontSizes } from 'utils/styleUtils';

// resources
import GetIdeasCount, { GetIdeasCountChildProps } from 'resources/GetIdeasCount';

// components
import CountBadge from 'components/UI/CountBadge';

const size = 21;
const padding = 4;

const Container = styled.div`
  display: inline-block;
  display: flex;
  align-items: center;
`;

const ToggleContainer: any = styled.div`
  height: 100%;
  display: flex;
  align-items: center;

  ${(props: any) => props.disabled && css`
    opacity: 0.25;

    i,
    i:before {
      cursor: not-allowed;
    }
  `};

  ${(props: any) => props.checked && css`
    i {
      padding-right: ${padding}px !important;
      padding-left: ${size}px !important;
      background: ${colors.clGreen} !important;
    }
  `};

  input {
    display: none;
  }

  i {
    display: inline-block;
    cursor: pointer;
    padding: ${padding}px;
    padding-right: ${size}px;
    transition: all ease 0.15s;
    border-radius: ${size + padding}px;
    background: #ccc;
    transform: translate3d(0, 0, 0);

    &:before {
      display: block;
      content: '';
      width: ${size}px;
      height: ${size}px;
      border-radius: ${size}px;
      background: #fff;
    }
  }
`;

const StyledLabel = styled.label`
  font-size: ${fontSizes.small}px;
  font-weight: 400;
  line-height: 20px;
  padding-left: 10px;
  padding-right: 10px;
  cursor: pointer;
  display: flex;
  align-items: center;
`;

interface InputProps {
  value: boolean;
  onChange: () => void;
  assignee: string;
}

interface DataProps {
  feedbackNeededCount: GetIdeasCountChildProps;
}

interface Props extends InputProps, DataProps {}

type State = {};

export class FeedbackToggle extends React.PureComponent<Props, State> {
  handleOnClick = () => {
    this.props.onChange();
  }

  render() {
    const { value, feedbackNeededCount } = this.props;

    return (
      <Container className="feedback_needed_filter_toggle">
        <StyledLabel onClick={this.handleOnClick}>
          <FormattedMessage {...messages.anyFeedbackStatus} />
        </StyledLabel>
        <ToggleContainer onClick={this.handleOnClick} checked={value}>
          <input type="checkbox" role="checkbox" aria-checked={value}/>
          <i />
        </ToggleContainer>
        <StyledLabel onClick={this.handleOnClick}>
          <FormattedMessage {...messages.needFeedback} />
          {!isNilOrError(feedbackNeededCount.count) && <CountBadge count={feedbackNeededCount.count}/>}
        </StyledLabel>
      </Container>
    );
  }
}

const Data = adopt({
  ideasCount: ({ assignee, render }) => (
    <GetIdeasCount feedbackNeeded={true} assignee={assignee === 'all' ? undefined : assignee}>
      {render}
    </GetIdeasCount>
  )
});

export default (inputProps: InputProps) => (
  <Data {...inputProps}>
    {dataProps => <FeedbackToggle {...inputProps} feedbackNeededCount={dataProps.ideasCount}/>}
  </Data>
);
