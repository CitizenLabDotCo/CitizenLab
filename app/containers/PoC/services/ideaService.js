import { API_PATH } from 'containers/App/constants';
import Streams from '../utils/streams';
import Rx from 'rxjs/Rx';
import _ from 'lodash';

export function observeSingleIdea(id, localProperties = false) {
  return Streams.create(`${API_PATH}/ideas/${id}`, null, localProperties);
}

export function observeMultipleIdeas(queryParameters = null, localProperties = false) {
  return Streams.create(`${API_PATH}/ideas`, queryParameters, localProperties);
}

export function observeIdeaRelationShips(idea) {
  const { relationships } = idea;

  const topicIds = _(relationships.topics.data).map((topic) => topic.id).value();
  const areaIds = _(relationships.areas.data).map((area) => area.id).value();
  const imageIds = _(relationships.idea_images.data).map((ideaImage) => ideaImage.id).value();
  const authorId = (relationships.author.data ? relationships.author.data.id : null);
  const projectId = (relationships.author.data ? relationships.project.data.id : null);

  const topicsStreams = (topicIds.length > 0 ? Rx.Observable.combineLatest(
    topicIds.map((topicId) => Streams.create(`${API_PATH}/topics/${topicId}`).observable),
    (...args) => _.flatten(args)
  ) : Rx.Observable.of(null));

  const areasStreams = (areaIds.length > 0 ? Rx.Observable.combineLatest(
    areaIds.map((areaId) => Streams.create(`${API_PATH}/areas/${areaId}`).observable),
    (...args) => _.flatten(args)
  ) : Rx.Observable.of(null));

  const imagesStreams = (imageIds.length > 0 ? Rx.Observable.combineLatest(
    imageIds.map((imageId) => Streams.create(`${API_PATH}/ideas/${idea.id}/images/${imageId}`).observable),
    (...args) => _.flatten(args)
  ) : Rx.Observable.of(null));

  const authorStream = (authorId ? Streams.create(`${API_PATH}/users/${authorId}`).observable : Rx.Observable.of(null));

  const projectStream = (projectId ? Streams.create(`${API_PATH}/projects/${projectId}`).observable : Rx.Observable.of(null));

  return Rx.Observable.combineLatest(
    topicsStreams,
    areasStreams,
    imagesStreams,
    authorStream,
    projectStream,
    (s1, s2, s3, s4, s5) => {
      return {
        topics: s1,
        areas: s2,
        images: s3,
        author: s4,
        project: s5,
      };
    }
  );
}

export function toggleIdea(observers, ideaID) {
  const action = (ideas) => ideas.map((idea) => {
    if (idea.id === ideaID) {
      return {
        ...idea,
        selected: !idea.selected,
      };
    }

    if (idea.id !== ideaID && idea.selected) {
      return {
        ...idea,
        selected: false,
      };
    }

    return idea;
  });

  observers.forEach((observer) => observer.next(action));
}
