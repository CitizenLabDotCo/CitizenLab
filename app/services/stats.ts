import { API_PATH } from 'containers/App/constants';
import streams, { IStreamParams } from 'utils/streams';
import { Multiloc } from 'typings';

const apiEndpoint = `${API_PATH}/stats`;

// Ideas
export interface IIdeasByTime {
  series: {
    ideas: {
      [key: string]: number;
    }
  };
}

export interface IIdeasByTopic {
  series: {
    ideas: {
      [key: string]: number;
    }
  };
  topics: {
    [key: string]: {
      title_multiloc: Multiloc
    }
  };
}

export interface IIdeasByProject {
  series: {
    ideas: {
      [key: string]: number;
    }
  };
  projects: {
    [key: string]: {
      title_multiloc: Multiloc
    }
  };
}

export function ideasByTimeStream(streamParams: IStreamParams | null = null) {
  return streams.get<IIdeasByTime>({ apiEndpoint: `${apiEndpoint}/ideas_by_time`, ...streamParams });
}

export function ideasByTimeCumulativeStream(streamParams: IStreamParams | null = null) {
  return streams.get<IIdeasByTime>({ apiEndpoint: `${apiEndpoint}/ideas_by_time_cumulative`, ...streamParams });
}

export function ideasByTopicStream(streamParams: IStreamParams | null = null) {
  return streams.get<IIdeasByTopic>({ apiEndpoint: `${apiEndpoint}/ideas_by_topic`, ...streamParams });
}

export function ideasByProjectStream(streamParams: IStreamParams | null = null) {
  return streams.get<IIdeasByProject>({ apiEndpoint: `${apiEndpoint}/ideas_by_project`, ...streamParams });
}

// Users
export interface IUsersByGender {
  series: {
    users: {
      [key: string]: number;
    }
  };
}

export interface IUsersCount {
  count: number;
}

export interface IUsersByBirthyear {
  series: {
    users: {
      [key: string]: number;
    }
  };
}

export interface IUsersByTime {
  series: {
    users: {
      [key: string]: number;
    }
  };
}

export function usersByGenderStream(streamParams: IStreamParams | null = null) {
  return streams.get<IUsersByGender>({ apiEndpoint: `${apiEndpoint}/users_by_gender`, ...streamParams });
}

export function usersCount(streamParams: IStreamParams | null = null) {
  return streams.get<IUsersCount>({ apiEndpoint: `${apiEndpoint}/users_count`, ...streamParams });
}

export function usersByBirthyearStream(streamParams: IStreamParams | null = null) {
  return streams.get<IUsersByBirthyear>({ apiEndpoint: `${apiEndpoint}/users_by_birthyear`, ...streamParams });
}

export function usersByTimeStream(streamParams: IStreamParams | null = null) {
  return streams.get<IUsersByTime>({ apiEndpoint: `${apiEndpoint}/users_by_time`, ...streamParams });
}

export function usersByTimeCumulativeStream(streamParams: IStreamParams | null = null) {
  return streams.get<IUsersByTime>({ apiEndpoint: `${apiEndpoint}/users_by_time_cumulative`, ...streamParams });
}

export function activeUsersByTimeStream(streamParams: IStreamParams | null = null) {
  return streams.get<IUsersByTime>({ apiEndpoint: `${apiEndpoint}/active_users_by_time`, ...streamParams });
}

// Comments
export interface ICommentsByTime {
  series: {
    comments: {
      [key: string]: number;
    }
  };
}

export interface ICommentsByTopic {
  series: {
    comments: {
      [key: string]: number;
    }
  };
  topics: {
    [key: string]: {
      title_multiloc: Multiloc
    }
  };
}

export interface ICommentsByProject {
  series: {
    comments: {
      [key: string]: number;
    }
  };
  projects: {
    [key: string]: {
      title_multiloc: Multiloc
    }
  };
}

export function commentsByTimeCumulativeStream(streamParams: IStreamParams | null = null) {
  return streams.get<ICommentsByTime>({ apiEndpoint: `${apiEndpoint}/comments_by_time_cumulative`, ...streamParams });
}

export function commentsByTopicStream(streamParams: IStreamParams | null = null) {
  return streams.get<ICommentsByTopic>({ apiEndpoint: `${apiEndpoint}/comments_by_topic`, ...streamParams });
}

export function commentsByProjectStream(streamParams: IStreamParams | null = null) {
  return streams.get<ICommentsByProject>({ apiEndpoint: `${apiEndpoint}/comments_by_project`, ...streamParams });
}

// Votes
export interface IVotesByTime {
  series: {
    votes: {
      [key: string]: number;
    }
  };
}

export interface IVotesByTopic {
  series: {
    votes: {
      [key: string]: number;
    }
  };
  topics: {
    [key: string]: {
      title_multiloc: Multiloc
    }
  };
}

export interface IVotesByProject {
  series: {
    votes: {
      [key: string]: number;
    }
  };
  projects: {
    [key: string]: {
      title_multiloc: Multiloc
    }
  };
}

interface IGenderCounts {
  male?: number;
  female?: number;
  unspecified?: number;
  _blank?: number;
}

export interface IVotesByGender {
  series: {
    up: IGenderCounts;
    down: IGenderCounts;
    total: IGenderCounts;
  };
}

export interface IVotesByTimeCumulative {
  series: {
    up: { [key: string]: number };
    down: { [key: string]: number };
    total: { [key: string]: number };
  };
}

export interface IVotesByBirthyear {
  series: {
    up: { [key: number]: number };
    down: { [key: number]: number };
    total: { [key: number]: number };
  };
}

export interface IVotesByDomicile {
  series: {
    up: { [key: string]: number };
    down: { [key: string]: number };
    total: { [key: string]: number };
  };
}

export function votesByTimeCumulativeStream(streamParams: IStreamParams | null = null) {
  return streams.get<IVotesByTimeCumulative>({ apiEndpoint: `${apiEndpoint}/votes_by_time_cumulative`, ...streamParams });
}

export function votesByTopicStream(streamParams: IStreamParams | null = null) {
  return streams.get<IVotesByTopic>({ apiEndpoint: `${apiEndpoint}/votes_by_topic`, ...streamParams });
}

export function votesByProjectStream(streamParams: IStreamParams | null = null) {
  return streams.get<IVotesByProject>({ apiEndpoint: `${apiEndpoint}/votes_by_project`, ...streamParams });
}

export function votesByGenderStream(streamParams: IStreamParams | null = null) {
  return streams.get<IVotesByGender>({ apiEndpoint: `${apiEndpoint}/votes_by_gender`, ...streamParams });
}

export function votesByBirthyearStream(streamParams: IStreamParams | null = null) {
  return streams.get<IVotesByBirthyear>({ apiEndpoint: `${apiEndpoint}/votes_by_birthyear`, ...streamParams });
}

export function votesByDomicileStream(streamParams: IStreamParams | null = null) {
  return streams.get<IVotesByDomicile>({ apiEndpoint: `${apiEndpoint}/votes_by_domicile`, ...streamParams });
}
