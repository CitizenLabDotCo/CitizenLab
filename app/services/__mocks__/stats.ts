import { BehaviorSubject } from 'rxjs';
import { IUsersByRegistrationField, IUsersByTime, IIdeasCount } from 'services/stats';
import { IStreamParams } from 'utils/streams';

let mockIdeasCount: IIdeasCount | null = null;

export const __setMockIdeasCount = (ideasCount: IIdeasCount) => {
  mockIdeasCount = ideasCount;
};

export const ideasCount = jest.fn((_ideasCount) => {
  const observable = new BehaviorSubject(mockIdeasCount);
  return {
    observable
  };
});

// usersByRegFieldStream
let mockUsersByRegFieldsVariable: IUsersByRegistrationField | null = null;

export const __setMockUsersByRegFields = (data: IUsersByRegistrationField) => {
  mockUsersByRegFieldsVariable = data;
};

export const usersByRegFieldStream = jest.fn((_streamParams: IStreamParams, _regFieldId) => {
  const observable = new BehaviorSubject(mockUsersByRegFieldsVariable);
  return {
    observable
  };
});

export const mockUsersByRegFields = {
  series: {
    users: {
      active_politician: 1,
      no: 7,
      retired_politician: 2,
      _blank: 11
    }
  },
  options: {
    active_politician: {
      title_multiloc: {
        en: 'Active politician'
      }
    },
    retired_politician: {
      title_multiloc: {
        en: 'Retired politician'
      }
    },
    no: {
      title_multiloc: {
        en: 'No'
      }
    }
  }
} as IUsersByRegistrationField;

// usersByTimeCumulativeStream
let mockUsersByTimeCumulativeStreamVariable: IUsersByTime | null = null;

export const __setMockUsersByTimeCumulativeStream = (data: IUsersByTime) => {
  mockUsersByTimeCumulativeStreamVariable = data;
};

export const usersByTimeCumulativeStream = jest.fn((_streamParams: IStreamParams) => {
  const observable = new BehaviorSubject(mockUsersByTimeCumulativeStreamVariable);
  return {
    observable
  };
});

export const mockUsersByTimeCumulative = {
  series: {
    users: {
      '2018-11-01': 1,
      '2018-11-02': 1,
      '2018-11-03': 1,
      '2018-11-04': 1,
      '2018-11-05': 1,
      '2018-11-06': 1,
      '2018-11-07': 1,
      '2018-11-08': 1,
      '2018-11-09': 1,
      '2018-11-10': 1,
      '2018-11-11': 6,
      '2018-11-12': 6,
      '2018-11-13': 6,
      '2018-11-14': 6,
      '2018-11-15': 6,
      '2018-11-16': 6,
      '2018-11-17': 6,
      '2018-11-18': 6,
      '2018-11-19': 6,
      '2018-11-20': 6,
      '2018-11-21': 6,
      '2018-11-22': 6,
      '2018-11-23': 6,
      '2018-11-24': 6,
      '2018-11-25': 6,
      '2018-11-26': 10,
      '2018-11-27': 10,
      '2018-11-28': 10,
      '2018-11-29': 10,
      '2018-11-30': 10
    }
  }
};
