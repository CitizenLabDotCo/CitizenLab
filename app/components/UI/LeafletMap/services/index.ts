import {
  setup,
  init,
  addTileLayer,
  addOnClickHandler,
  changeView,
  addOnMoveHandler,
  addOnZoomHandler,
  refitBounds,
} from './setup';

import { addLayersControl, removeControl } from './controls';
import { addLayers, removeLayers, removeLayer } from './layers';
import { addMarkersToMap, addClusterGroup } from './markers';

export default {
  setup,
  init,
  addTileLayer,
  addOnClickHandler,
  changeView,
  addOnMoveHandler,
  addOnZoomHandler,
  refitBounds,
  addLayersControl,
  removeControl,
  addLayers,
  removeLayers,
  removeLayer,
  addMarkersToMap,
  addClusterGroup,
};
