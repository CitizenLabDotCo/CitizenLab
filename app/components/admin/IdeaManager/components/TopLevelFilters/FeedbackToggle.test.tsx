import React from 'react';
import { shallow } from 'enzyme';

import { FeedbackToggle } from './FeedbackToggle';
import { makeIdeasCount } from 'services/stats';

jest.mock('services/stats');
jest.mock('utils/cl-intl');

import 'jest-styled-components';

describe('<FeedbackToggle />', () => {
  let onChange: jest.Mock;
  beforeEach(() => {
    onChange = jest.fn();
  });

  it('renders correctly unchecked', () => {
    const ideasCount = makeIdeasCount(6);

    const wrapper = shallow(
      <FeedbackToggle
        assignee="me"
        feedbackNeededCount={ideasCount}
        value={false}
        onChange={onChange}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly checked', () => {
    const ideasCount = makeIdeasCount(6);

    const wrapper = shallow(
      <FeedbackToggle
        assignee="me"
        feedbackNeededCount={ideasCount}
        value={true}
        onChange={onChange}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calls onChange when clicked', () => {
    const ideasCount = makeIdeasCount(6);

    const wrapper = shallow(
      <FeedbackToggle
        assignee="me"
        feedbackNeededCount={ideasCount}
        value={false}
        onChange={onChange}
      />
    );
    wrapper.find('FeedbackToggle__ToggleContainer').simulate('click', { preventDefault() { }, stopPropagation() { } });

    expect(onChange).toHaveBeenCalled();
  });
});
