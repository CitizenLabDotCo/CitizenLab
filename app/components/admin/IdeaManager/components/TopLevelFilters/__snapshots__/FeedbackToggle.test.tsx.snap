// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeedbackToggle /> renders correctly checked 1`] = `
<FeedbackToggle__Container
  className="feedback_needed_filter_toggle"
  id="e2e-feedback_needed_filter_toggle"
>
  <FeedbackToggle__StyledLabel
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="All ideas"
      id="app.components.admin.IdeaManager.anyFeedbackStatus"
    />
  </FeedbackToggle__StyledLabel>
  <FeedbackToggle__ToggleContainer
    checked={true}
    onClick={[Function]}
  >
    <input
      aria-checked={true}
      role="checkbox"
      type="checkbox"
    />
    <i />
  </FeedbackToggle__ToggleContainer>
  <FeedbackToggle__StyledLabel
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Need Feedback"
      id="app.components.admin.IdeaManager.needFeedback"
    />
    <Component
      count={6}
    />
  </FeedbackToggle__StyledLabel>
</FeedbackToggle__Container>
`;

exports[`<FeedbackToggle /> renders correctly unchecked 1`] = `
<FeedbackToggle__Container
  className="feedback_needed_filter_toggle"
  id="e2e-feedback_needed_filter_toggle"
>
  <FeedbackToggle__StyledLabel
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="All ideas"
      id="app.components.admin.IdeaManager.anyFeedbackStatus"
    />
  </FeedbackToggle__StyledLabel>
  <FeedbackToggle__ToggleContainer
    checked={false}
    onClick={[Function]}
  >
    <input
      aria-checked={false}
      role="checkbox"
      type="checkbox"
    />
    <i />
  </FeedbackToggle__ToggleContainer>
  <FeedbackToggle__StyledLabel
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Need Feedback"
      id="app.components.admin.IdeaManager.needFeedback"
    />
    <Component
      count={6}
    />
  </FeedbackToggle__StyledLabel>
</FeedbackToggle__Container>
`;
